Spigot-Server-Patches/0532-PiglinBarterEvent-and-Currency-API.patch

```diff
@@ -0,0 +1,257 @@
+             c(entitypiglin, itemstack);
+             d((EntityLiving) entitypiglin);
+         } else if (c(item) && !A(entitypiglin)) {
+@@ -190,9 +193,19 @@ public class PiglinAI {
+         boolean flag1;
+ 
+         if (entitypiglin.eM()) {
+-            flag1 = b(itemstack.getItem());
++            flag1 = isBarterCurrency(itemstack.getItem(), entitypiglin); // Paper
+             if (flag && flag1) {
+-                a(entitypiglin, m(entitypiglin));
++                // Paper start - piglin barter
++                List<ItemStack> results = m(entitypiglin);
++                List<org.bukkit.inventory.ItemStack> bukkitResults = com.google.common.collect.Lists.newArrayList();
++                results.forEach(r -> bukkitResults.add(r.asBukkitMirror()));
++                com.destroystokyo.paper.event.entity.PiglinBarterEvent event = new com.destroystokyo.paper.event.entity.PiglinBarterEvent((org.bukkit.entity.Piglin) entitypiglin.getBukkitLivingEntity(), entitypiglin.itemThrower != null ? org.bukkit.Bukkit.getPlayer(entitypiglin.itemThrower) : null, CraftMagicNumbers.getMaterial(itemstack.getItem()) ,bukkitResults);
++                if (event.callEvent()) {
++                    results.clear();
++                    event.getResults().forEach(bResult -> results.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(bResult)));
++                    a(entitypiglin, results);
++                }
++                // Paper end
+             } else if (!flag1) {
+                 boolean flag2 = entitypiglin.g(itemstack);
+ 
+@@ -275,22 +288,26 @@ public class PiglinAI {
+         return entityliving1.getEntityType() != EntityTypes.HOGLIN ? false : (new Random(entityliving.world.getTime())).nextFloat() < 0.1F;
+     }
+ 
+-    protected static boolean a(EntityPiglin entitypiglin, ItemStack itemstack) {
++    protected static boolean a(EntityPiglin entitypiglin, ItemStack itemstack) { // Paper - wants to pick up
+         Item item = itemstack.getItem();
+ 
+         if (item.a((Tag) TagsItem.PIGLIN_REPELLENTS)) {
+             return false;
+         } else if (D(entitypiglin) && entitypiglin.getBehaviorController().hasMemory(MemoryModuleType.ATTACK_TARGET)) {
+             return false;
+-        } else if (b(item)) {
++        } else if (isBarterCurrency(item, entitypiglin)) {
+             return F(entitypiglin);
+         } else {
+             boolean flag = entitypiglin.l(itemstack);
+ 
+-            return item == Items.GOLD_NUGGET ? flag : (c(item) ? !A(entitypiglin) && flag : (!a(item) ? entitypiglin.o(itemstack) : F(entitypiglin) && flag));
++            return item == Items.GOLD_NUGGET ? flag : (c(item) ? !A(entitypiglin) && flag : (!isLovedItem(item, entitypiglin) ? entitypiglin.o(itemstack) : F(entitypiglin) && flag));
+         }
+     }
+-
++    // Paper start
++    protected static boolean isLovedItem(Item item, EntityPiglin entityPiglin) {
++        return a(item) || entityPiglin.getBarterCurrency().contains(item);
```

Comment:
Maybe single line this method so it looks more like an OBFHELPER. idk

Reply:
yeah, I didn't cause its not technically and OBFHELPER